server/server.go

```diff
@@ -177,10 +185,54 @@ func (s *Server) authUser(c ssh.ConnMetadata, password []byte) (*ssh.Permissions
 	// check the user exists and has matching password
 	n := c.User()
 	user, found := s.users.Get(n)
-	if !found || user.Pass != string(password) {
-		s.Debugf("Login failed for user: %s", n)
-		return nil, errors.New("Invalid authentication for username: %s")
+	if (found) {
+	// User found
+
+		if (s.config.LdapConfigFile == "") {
+			// LDAP disabled
+
+			if (string(password) == "") {
+				// Empty password
+
+				s.Debugf("Empty passwords are not allowed. User = %s", n)
+				return nil, errors.New("Invalid authentication for username: %s")
+			} else if (user.Pass == string(password)) {
+				// Password authentication successful.
+
+			} else {
+				// Password authentication failed.
+
+				s.Debugf("Could not authenticate user: %s", n)
+				return nil, errors.New("Invalid authentication for username: %s")
+			}
+		} else {
+			// LDAP enabled
+
+			if (string(password) == "") {
+				// Empty password
+
+				s.Debugf("Empty passwords are not allowed. User = %s", n)
+				return nil, errors.New("Invalid authentication for username: %s")
+			} else if (user.Pass == string(password)) {
+				// Password authentication successful.
+
+			} else if (settings.LdapAuthUser(user,password,s.config.LdapConfig) == nil) {
+				// LDAP authentication successful.
+
+			} else {
+				// Password authentication & LDAP authentication failed.
+
+				s.Debugf("Could not authenticate user: %s", n)
+				return nil, errors.New("Invalid authentication for username: %s")
+			}
+		}
+	} else {
+	// User not found
+
+	s.Debugf("User not found: %s", n)
+	return nil, errors.New("Invalid authentication for username: %s")
 	}
```

Comment:
@kblanckaert
instead of
	if (!ldap_enabled && string(password) == "") {
		// Empty password

		s.Debugf("Empty passwords are not allowed. User = %s", n)
		return nil, errors.New("Invalid authentication for username: %s")
 I would rather do this
	if (string(password) == "") {
		// Empty password

		s.Debugf("Empty passwords are not allowed. User = %s", n)
		return nil, errors.New("Invalid authentication for username: %s")

giving that empty password is not allowed anyways

Reply:
@kblanckaert 
and the stanza 
	} else if (settings.LdapAuthUser(user,password,s.config.LdapConfig) == nil) {
		// LDAP authentication successful.
should be replaced by 
	} else if (ldap_enabled && settings.LdapAuthUser(user,password,s.config.LdapConfig) == nil) {
		// LDAP authentication successful.

to avoid failure

But to be honest I a not sure we are gaining any simplification vs what is currently proposed in the PR

Reply:
I've added that to avoid logging in with an empty password when there is no LDAP, but the same can be said for every placeholder character. Password authentication is still checked before LDAP, so you could log in by using that placeholder character.
Another option is checking LDAP authentication first (if enabled)

The 'ldap_enabled' should indeed be added to third _else if_ (especially when swapping the password & ldap auth statements)